<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Module 3: Generalized Linear Models | Analysis Using R</title>
  <meta name="description" content="Principles of exploratory data analysis, RNAseq differential expression, and generalized linear models" />
  <meta name="generator" content="bookdown 0.24 and GitBook 2.6.7" />

  <meta property="og:title" content="Module 3: Generalized Linear Models | Analysis Using R" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Principles of exploratory data analysis, RNAseq differential expression, and generalized linear models" />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Module 3: Generalized Linear Models | Analysis Using R" />
  
  <meta name="twitter:description" content="Principles of exploratory data analysis, RNAseq differential expression, and generalized linear models" />
  

<meta name="author" content="Instructors: Shraddha Pai, Chaitra Sarathy" />



  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  <link rel="shortcut icon" href="images/favicon.ico" type="image/x-icon" />
<link rel="prev" href="module-2-exercise.html"/>
<link rel="next" href="module-3-exercise.html"/>
<script src="libs/header-attrs-2.11/header-attrs.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<img src="images/bioinfologo_big.png">
<li><a href="./">2023 CBW Analysis Using R Workshop</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#meet-your-faculty"><i class="fa fa-check"></i>Meet your Faculty</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#pre-workshop-materials-and-laptop-setup-instructions"><i class="fa fa-check"></i>Pre-workshop Materials and Laptop Setup Instructions</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#laptop-setup-instructions"><i class="fa fa-check"></i>Laptop Setup Instructions</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#r-packages"><i class="fa fa-check"></i>R packages</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="lecture-slides.html"><a href="lecture-slides.html"><i class="fa fa-check"></i>Lecture slides</a></li>
<li class="chapter" data-level="" data-path="module-1-exploratory-data-analysis-and-clustering.html"><a href="module-1-exploratory-data-analysis-and-clustering.html"><i class="fa fa-check"></i>Module 1: Exploratory Data Analysis and Clustering</a>
<ul>
<li><a href="module-1-exploratory-data-analysis-and-clustering.html#load-mouse-data">Load <code>mouse</code> data</a></li>
<li><a href="module-1-exploratory-data-analysis-and-clustering.html#rcolorbrewer-for-colour-palettes"><code>RColorBrewer</code> for colour palettes</a></li>
<li class="chapter" data-level="" data-path="module-1-exploratory-data-analysis-and-clustering.html"><a href="module-1-exploratory-data-analysis-and-clustering.html#correlations-distances-and-clustering"><i class="fa fa-check"></i>Correlations, distances, and clustering</a></li>
<li class="chapter" data-level="" data-path="module-1-exploratory-data-analysis-and-clustering.html"><a href="module-1-exploratory-data-analysis-and-clustering.html#hierarchical-clustering"><i class="fa fa-check"></i>Hierarchical clustering</a></li>
<li class="chapter" data-level="" data-path="module-1-exploratory-data-analysis-and-clustering.html"><a href="module-1-exploratory-data-analysis-and-clustering.html#k-means-clustering"><i class="fa fa-check"></i>K-means clustering</a></li>
<li><a href="module-1-exploratory-data-analysis-and-clustering.html#using-clvalid-to-determine-number-of-clusters">Using <code>clValid</code> to determine number of clusters</a></li>
<li class="chapter" data-level="" data-path="module-1-exploratory-data-analysis-and-clustering.html"><a href="module-1-exploratory-data-analysis-and-clustering.html#exercise"><i class="fa fa-check"></i>Exercise</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-1-exercise-results.html"><a href="module-1-exercise-results.html"><i class="fa fa-check"></i>Module 1 Exercise Results</a></li>
<li class="chapter" data-level="" data-path="module-2-dimensionality-reduction.html"><a href="module-2-dimensionality-reduction.html"><i class="fa fa-check"></i>Module 2: Dimensionality reduction</a>
<ul>
<li class="chapter" data-level="" data-path="module-2-dimensionality-reduction.html"><a href="module-2-dimensionality-reduction.html#exercise-1"><i class="fa fa-check"></i>Exercise</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-2-exercise.html"><a href="module-2-exercise.html"><i class="fa fa-check"></i>Module 2 exercise</a></li>
<li class="chapter" data-level="" data-path="module-3-generalized-linear-models.html"><a href="module-3-generalized-linear-models.html"><i class="fa fa-check"></i>Module 3: Generalized Linear Models</a>
<ul>
<li class="chapter" data-level="" data-path="module-3-generalized-linear-models.html"><a href="module-3-generalized-linear-models.html#read-data-from-files-and-explore-variable-distribution"><i class="fa fa-check"></i>Read data from files and explore variable distribution</a></li>
<li class="chapter" data-level="" data-path="module-3-generalized-linear-models.html"><a href="module-3-generalized-linear-models.html#explore-missing-data"><i class="fa fa-check"></i>Explore missing data</a></li>
<li><a href="module-3-generalized-linear-models.html#create-plots-with-ggplot2-to-explore-variable-relationships">Create plots with <code>ggplot2</code> to explore variable relationships</a></li>
<li><a href="module-3-generalized-linear-models.html#fit-binary-response-variable-using-glm-and-logistic-regression">Fit binary response variable using <code>glm()</code> and logistic regression</a></li>
<li class="chapter" data-level="" data-path="module-3-generalized-linear-models.html"><a href="module-3-generalized-linear-models.html#exercise-2"><i class="fa fa-check"></i>Exercise</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-3-exercise.html"><a href="module-3-exercise.html"><i class="fa fa-check"></i>Module 3 exercise</a></li>
<li class="chapter" data-level="" data-path="module-4-finding-differentially-expressed-genes-with-rna-seq.html"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html"><i class="fa fa-check"></i>Module 4: Finding differentially expressed genes with RNA-seq</a>
<ul>
<li class="chapter" data-level="" data-path="module-4-finding-differentially-expressed-genes-with-rna-seq.html"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#mini-introduction-to-bioconductor"><i class="fa fa-check"></i>Mini introduction to BioConductor</a></li>
<li><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#fetch-breast-cancer-data-using-curatedtcgadata">Fetch breast cancer data using <code>curatedTCGAData</code></a>
<ul>
<li class="chapter" data-level="" data-path="module-4-finding-differentially-expressed-genes-with-rna-seq.html"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#prepare-data-for-differential-expression-analysis"><i class="fa fa-check"></i>Prepare data for differential expression analysis</a></li>
</ul></li>
<li><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#differential-expression-analysis-with-edger">Differential expression analysis with <code>edgeR</code></a></li>
<li class="chapter" data-level="" data-path="module-4-finding-differentially-expressed-genes-with-rna-seq.html"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#exercise-3"><i class="fa fa-check"></i>Exercise</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-4---exercises.html"><a href="module-4---exercises.html"><i class="fa fa-check"></i>Module 4 - exercises</a></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Analysis Using R</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="module-3-generalized-linear-models" class="section level1 unnumbered">
<h1>Module 3: Generalized Linear Models</h1>
<p>In this module we’re going to cover:</p>
<ul>
<li>Reading data from files and evaluating missingness</li>
<li>Creating publication-quality plots with <code>ggplot2</code></li>
<li>Fitting models with binary outcomes using generalized linear models</li>
</ul>
<div id="read-data-from-files-and-explore-variable-distribution" class="section level2 unnumbered">
<h2>Read data from files and explore variable distribution</h2>
<p>We’re going to use a popular dataset for data analysis, pertaining to survival of passengers aboard the Titanic.</p>
<p>Let’s read in the data from a file:</p>
<div class="sourceCode" id="cb78"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb78-1"><a href="module-3-generalized-linear-models.html#cb78-1" aria-hidden="true" tabindex="-1"></a>dat <span class="ot">&lt;-</span> <span class="fu">read.delim</span>(</span>
<span id="cb78-2"><a href="module-3-generalized-linear-models.html#cb78-2" aria-hidden="true" tabindex="-1"></a>    <span class="st">&quot;data/titanic.csv&quot;</span>,</span>
<span id="cb78-3"><a href="module-3-generalized-linear-models.html#cb78-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">sep=</span><span class="st">&quot;,&quot;</span> <span class="co"># indicate the column separator</span></span>
<span id="cb78-4"><a href="module-3-generalized-linear-models.html#cb78-4" aria-hidden="true" tabindex="-1"></a>    ) </span></code></pre></div>
<p>Examine the columns:</p>
<div class="sourceCode" id="cb79"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb79-1"><a href="module-3-generalized-linear-models.html#cb79-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(dat)</span></code></pre></div>
<pre><code>##   PassengerId Survived Pclass                                                Name
## 1           1        0      3                             Braund, Mr. Owen Harris
## 2           2        1      1 Cumings, Mrs. John Bradley (Florence Briggs Thayer)
## 3           3        1      3                              Heikkinen, Miss. Laina
## 4           4        1      1        Futrelle, Mrs. Jacques Heath (Lily May Peel)
## 5           5        0      3                            Allen, Mr. William Henry
## 6           6        0      3                                    Moran, Mr. James
##      Sex Age SibSp Parch           Ticket    Fare Cabin Embarked
## 1   male  22     1     0        A/5 21171  7.2500              S
## 2 female  38     1     0         PC 17599 71.2833   C85        C
## 3 female  26     0     0 STON/O2. 3101282  7.9250              S
## 4 female  35     1     0           113803 53.1000  C123        S
## 5   male  35     0     0           373450  8.0500              S
## 6   male  NA     0     0           330877  8.4583              Q</code></pre>
<p>Some of the columns are categorical, use <code>table()</code> to look at the tallies:
Examine the columns:</p>
<div class="sourceCode" id="cb81"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb81-1"><a href="module-3-generalized-linear-models.html#cb81-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(dat<span class="sc">$</span>Survived)</span></code></pre></div>
<pre><code>## 
##   0   1 
## 549 342</code></pre>
<div class="sourceCode" id="cb83"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb83-1"><a href="module-3-generalized-linear-models.html#cb83-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(dat<span class="sc">$</span>Pclass)</span></code></pre></div>
<pre><code>## 
##   1   2   3 
## 216 184 491</code></pre>
<p>Use <code>summary()</code> to look at continuous-valued data:</p>
<div class="sourceCode" id="cb85"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb85-1"><a href="module-3-generalized-linear-models.html#cb85-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(dat<span class="sc">$</span>Age)</span></code></pre></div>
<pre><code>##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max.    NA&#39;s 
##    0.42   20.12   28.00   29.70   38.00   80.00     177</code></pre>
<p>Notice that there are 177 <code>NA</code> (missing) values for age.
Let’s visualize the missing data more systematically.</p>
</div>
<div id="explore-missing-data" class="section level2 unnumbered">
<h2>Explore missing data</h2>
<p>For this let’s use a little script that converts a table into black and white squares to visualize missing data. For this, install the <code>plotrix</code> package</p>
<div class="sourceCode" id="cb87"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb87-1"><a href="module-3-generalized-linear-models.html#cb87-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">&quot;plotrix&quot;</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>)) <span class="fu">install.packages</span>(<span class="st">&quot;plotrix&quot;</span>)</span>
<span id="cb87-2"><a href="module-3-generalized-linear-models.html#cb87-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb87-3"><a href="module-3-generalized-linear-models.html#cb87-3" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">require</span>(plotrix))</span>
<span id="cb87-4"><a href="module-3-generalized-linear-models.html#cb87-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb87-5"><a href="module-3-generalized-linear-models.html#cb87-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&#39; show data missingness as a chequered matrix</span></span>
<span id="cb87-6"><a href="module-3-generalized-linear-models.html#cb87-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&#39; </span></span>
<span id="cb87-7"><a href="module-3-generalized-linear-models.html#cb87-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&#39; @param x (matrix) data matrix.</span></span>
<span id="cb87-8"><a href="module-3-generalized-linear-models.html#cb87-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&#39; @param outFile (char) path to file for printing graph</span></span>
<span id="cb87-9"><a href="module-3-generalized-linear-models.html#cb87-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&#39; @param wd (numeric) width in inches</span></span>
<span id="cb87-10"><a href="module-3-generalized-linear-models.html#cb87-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&#39; @param ht (numeric) height in inches</span></span>
<span id="cb87-11"><a href="module-3-generalized-linear-models.html#cb87-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&#39; @return plots missingness matrix to file</span></span>
<span id="cb87-12"><a href="module-3-generalized-linear-models.html#cb87-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&#39; @import plotrix</span></span>
<span id="cb87-13"><a href="module-3-generalized-linear-models.html#cb87-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&#39; @export</span></span>
<span id="cb87-14"><a href="module-3-generalized-linear-models.html#cb87-14" aria-hidden="true" tabindex="-1"></a>plotMissMat <span class="ot">&lt;-</span> <span class="cf">function</span>(x,<span class="at">xlab=</span><span class="st">&quot;columns&quot;</span>,</span>
<span id="cb87-15"><a href="module-3-generalized-linear-models.html#cb87-15" aria-hidden="true" tabindex="-1"></a>        <span class="at">ylab=</span><span class="st">&quot;rows&quot;</span>,<span class="at">border=</span><span class="cn">NA</span>) {</span>
<span id="cb87-16"><a href="module-3-generalized-linear-models.html#cb87-16" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb87-17"><a href="module-3-generalized-linear-models.html#cb87-17" aria-hidden="true" tabindex="-1"></a>    x <span class="ot">&lt;-</span> <span class="sc">!</span><span class="fu">is.na</span>(x)</span>
<span id="cb87-18"><a href="module-3-generalized-linear-models.html#cb87-18" aria-hidden="true" tabindex="-1"></a>    <span class="fu">class</span>(x) <span class="ot">&lt;-</span> <span class="st">&quot;numeric&quot;</span></span>
<span id="cb87-19"><a href="module-3-generalized-linear-models.html#cb87-19" aria-hidden="true" tabindex="-1"></a>    <span class="fu">color2D.matplot</span>(x,<span class="at">show.values=</span><span class="cn">FALSE</span>,<span class="at">axes=</span><span class="cn">FALSE</span>,</span>
<span id="cb87-20"><a href="module-3-generalized-linear-models.html#cb87-20" aria-hidden="true" tabindex="-1"></a>        <span class="at">cs1=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">0</span>),<span class="at">cs2=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">0</span>),<span class="at">cs3=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">0</span>),<span class="at">border=</span>border,</span>
<span id="cb87-21"><a href="module-3-generalized-linear-models.html#cb87-21" aria-hidden="true" tabindex="-1"></a>        <span class="at">cex=</span><span class="fl">0.8</span>,</span>
<span id="cb87-22"><a href="module-3-generalized-linear-models.html#cb87-22" aria-hidden="true" tabindex="-1"></a>        <span class="at">xlab=</span>xlab,<span class="at">ylab=</span>ylab)</span>
<span id="cb87-23"><a href="module-3-generalized-linear-models.html#cb87-23" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<p>Let’s look at the missingness in the Titanic dataset. Missing data is shown as a white cell, and non-missing data is shown in black.</p>
<div class="sourceCode" id="cb88"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb88-1"><a href="module-3-generalized-linear-models.html#cb88-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotMissMat</span>(dat)</span></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/unnamed-chunk-57-1.png" width="672" /></p>
<p>We can see a column with many missing values. This is probably the “age” data.
Let’s count the number of missing values on a per-column level.</p>
<p>For this we combine <code>is.na()</code>, which returns a TRUE/FALSE value for NA values, and <code>colSums()</code> which adds up the TRUE values down the columns.</p>
<div class="sourceCode" id="cb89"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb89-1"><a href="module-3-generalized-linear-models.html#cb89-1" aria-hidden="true" tabindex="-1"></a><span class="fu">colSums</span>(<span class="fu">is.na</span>(dat))</span></code></pre></div>
<pre><code>## PassengerId    Survived      Pclass        Name         Sex         Age       SibSp 
##           0           0           0           0           0         177           0 
##       Parch      Ticket        Fare       Cabin    Embarked 
##           0           0           0           0           0</code></pre>
<p>This confirms that <code>Age</code> is the only column with missing data.</p>
<p>Now let’s explore the data using plots.</p>
</div>
<div id="create-plots-with-ggplot2-to-explore-variable-relationships" class="section level2 unnumbered">
<h2>Create plots with <code>ggplot2</code> to explore variable relationships</h2>
<p><code>ggplot2</code> is a popular plotting package that uses an additive paradigm to build plots. The <a href="">ggplot2 website</a> is a wealth of reference information, so here we just touch on the basics to get you started.</p>
<p>Anytime you need to generate a specific kind of plot, the website will most likely have documentation for how to achieve it.</p>
<p>Let’s start by creating a scatterplot. For this let’s load a dataset measuring statistics around quality of life in US states in the late 70’s:</p>
<div class="sourceCode" id="cb91"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb91-1"><a href="module-3-generalized-linear-models.html#cb91-1" aria-hidden="true" tabindex="-1"></a>state.x77 <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(state.x77)</span>
<span id="cb91-2"><a href="module-3-generalized-linear-models.html#cb91-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(state.x77)</span></code></pre></div>
<pre><code>##            Population Income Illiteracy Life Exp Murder HS Grad Frost   Area
## Alabama          3615   3624        2.1    69.05   15.1    41.3    20  50708
## Alaska            365   6315        1.5    69.31   11.3    66.7   152 566432
## Arizona          2212   4530        1.8    70.55    7.8    58.1    15 113417
## Arkansas         2110   3378        1.9    70.66   10.1    39.9    65  51945
## California      21198   5114        1.1    71.71   10.3    62.6    20 156361
## Colorado         2541   4884        0.7    72.06    6.8    63.9   166 103766</code></pre>
<p>Create a base plot using the <code>ggplot</code> function. Then “add” a scatterplot to it. Notice that the plot has been assigned to a variable named <code>p</code>.</p>
<p>This setup is standard for <code>ggplot2</code> and allows multiple visualizations to be applied to the same base plot.</p>
<div class="sourceCode" id="cb93"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb93-1"><a href="module-3-generalized-linear-models.html#cb93-1" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(ggplot2)</span>
<span id="cb93-2"><a href="module-3-generalized-linear-models.html#cb93-2" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(state.x77, <span class="fu">aes</span>(Illiteracy,Income))</span>
<span id="cb93-3"><a href="module-3-generalized-linear-models.html#cb93-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-4"><a href="module-3-generalized-linear-models.html#cb93-4" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">geom_point</span>() <span class="co"># scatter plot</span></span>
<span id="cb93-5"><a href="module-3-generalized-linear-models.html#cb93-5" aria-hidden="true" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/load-states-data-1.png" width="672" /></p>
<p>Now let’s add confidence intervals:</p>
<div class="sourceCode" id="cb94"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb94-1"><a href="module-3-generalized-linear-models.html#cb94-1" aria-hidden="true" tabindex="-1"></a>p <span class="sc">+</span> <span class="fu">geom_smooth</span>()</span></code></pre></div>
<pre><code>## `geom_smooth()` using method = &#39;loess&#39; and formula = &#39;y ~ x&#39;</code></pre>
<p><img src="bookdown-demo_files/figure-html/states-geomsmooth-1.png" width="672" /></p>
<p>It looks like there is a negative relationship between illiteracy and income.
We can confirm this by looking at correlation:</p>
<div class="sourceCode" id="cb96"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb96-1"><a href="module-3-generalized-linear-models.html#cb96-1" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> state.x77<span class="sc">$</span>Illiteracy</span>
<span id="cb96-2"><a href="module-3-generalized-linear-models.html#cb96-2" aria-hidden="true" tabindex="-1"></a>y <span class="ot">&lt;-</span> state.x77<span class="sc">$</span>Income</span>
<span id="cb96-3"><a href="module-3-generalized-linear-models.html#cb96-3" aria-hidden="true" tabindex="-1"></a><span class="fu">cor.test</span>(x,y)</span></code></pre></div>
<pre><code>## 
##  Pearson&#39;s product-moment correlation
## 
## data:  x and y
## t = -3.3668, df = 48, p-value = 0.001505
## alternative hypothesis: true correlation is not equal to 0
## 95 percent confidence interval:
##  -0.6378257 -0.1807128
## sample estimates:
##        cor 
## -0.4370752</code></pre>
<div class="sourceCode" id="cb98"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb98-1"><a href="module-3-generalized-linear-models.html#cb98-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cor.test</span>(x,y)<span class="sc">$</span>p.value</span></code></pre></div>
<pre><code>## [1] 0.001505073</code></pre>
<p>Let’s now examine categorical variables.
In the titanic set, let’s look at the fare paid based on passenger class:</p>
<div class="sourceCode" id="cb100"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb100-1"><a href="module-3-generalized-linear-models.html#cb100-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(dat)</span>
<span id="cb100-2"><a href="module-3-generalized-linear-models.html#cb100-2" aria-hidden="true" tabindex="-1"></a>p <span class="sc">+</span> <span class="fu">geom_boxplot</span>(</span>
<span id="cb100-3"><a href="module-3-generalized-linear-models.html#cb100-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(<span class="fu">factor</span>(Pclass), <span class="co"># make categorical</span></span>
<span id="cb100-4"><a href="module-3-generalized-linear-models.html#cb100-4" aria-hidden="true" tabindex="-1"></a>        Fare))</span></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/boxplot-1.png" width="672" /></p>
<p>We can use barplots to examine counts and proportions.
Let’s look at number of survivors, split by passenger class</p>
<div class="sourceCode" id="cb101"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb101-1"><a href="module-3-generalized-linear-models.html#cb101-1" aria-hidden="true" tabindex="-1"></a>p <span class="sc">+</span> <span class="fu">geom_bar</span>(</span>
<span id="cb101-2"><a href="module-3-generalized-linear-models.html#cb101-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(<span class="at">fill=</span><span class="fu">factor</span>(Survived), Pclass)</span>
<span id="cb101-3"><a href="module-3-generalized-linear-models.html#cb101-3" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/barplot-1.png" width="672" /></p>
<p>The plot above shows count data. Let’s convert this to proportions. We can see that the fraction of non-survivors in “Class 3” is high.</p>
<div class="sourceCode" id="cb102"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb102-1"><a href="module-3-generalized-linear-models.html#cb102-1" aria-hidden="true" tabindex="-1"></a>p <span class="sc">+</span> <span class="fu">geom_bar</span>(</span>
<span id="cb102-2"><a href="module-3-generalized-linear-models.html#cb102-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(<span class="at">fill=</span><span class="fu">factor</span>(Survived), Pclass),</span>
<span id="cb102-3"><a href="module-3-generalized-linear-models.html#cb102-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">position =</span> <span class="st">&quot;fill&quot;</span></span>
<span id="cb102-4"><a href="module-3-generalized-linear-models.html#cb102-4" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/barplot2%20-1.png" width="672" /></p>
<p>How about males versus females?</p>
<div class="sourceCode" id="cb103"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb103-1"><a href="module-3-generalized-linear-models.html#cb103-1" aria-hidden="true" tabindex="-1"></a>p <span class="sc">+</span> <span class="fu">geom_bar</span>(</span>
<span id="cb103-2"><a href="module-3-generalized-linear-models.html#cb103-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(<span class="at">fill=</span><span class="fu">factor</span>(Survived), Sex),</span>
<span id="cb103-3"><a href="module-3-generalized-linear-models.html#cb103-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">position =</span> <span class="st">&quot;fill&quot;</span></span>
<span id="cb103-4"><a href="module-3-generalized-linear-models.html#cb103-4" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/finish-titanic-1.png" width="672" /></p>
</div>
<div id="fit-binary-response-variable-using-glm-and-logistic-regression" class="section level2 unnumbered">
<h2>Fit binary response variable using <code>glm()</code> and logistic regression</h2>
<p>Let’s fit a model to a binary outcome. For this we load a dataset that measures physiological variables in a cohort of Pima Indians.</p>
<div class="sourceCode" id="cb104"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb104-1"><a href="module-3-generalized-linear-models.html#cb104-1" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(mlbench)</span></code></pre></div>
<pre><code>## Loading required package: mlbench</code></pre>
<pre><code>## Warning: package &#39;mlbench&#39; was built under R version 4.1.2</code></pre>
<div class="sourceCode" id="cb107"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb107-1"><a href="module-3-generalized-linear-models.html#cb107-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(PimaIndiansDiabetes2)</span>
<span id="cb107-2"><a href="module-3-generalized-linear-models.html#cb107-2" aria-hidden="true" tabindex="-1"></a><span class="co"># type ?PimaIndiansDiabetes2 to learn more about the dataset.</span></span>
<span id="cb107-3"><a href="module-3-generalized-linear-models.html#cb107-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb107-4"><a href="module-3-generalized-linear-models.html#cb107-4" aria-hidden="true" tabindex="-1"></a>dat <span class="ot">&lt;-</span> PimaIndiansDiabetes2</span>
<span id="cb107-5"><a href="module-3-generalized-linear-models.html#cb107-5" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(dat)</span></code></pre></div>
<pre><code>##   pregnant glucose pressure triceps insulin mass pedigree age diabetes
## 1        6     148       72      35      NA 33.6    0.627  50      pos
## 2        1      85       66      29      NA 26.6    0.351  31      neg
## 3        8     183       64      NA      NA 23.3    0.672  32      pos
## 4        1      89       66      23      94 28.1    0.167  21      neg
## 5        0     137       40      35     168 43.1    2.288  33      pos
## 6        5     116       74      NA      NA 25.6    0.201  30      neg</code></pre>
<p>Let’s look at the impact of blood glucose levels on diabetes diagnosis. First let’s make a scatterplot.</p>
<p>Could there be a relationship?</p>
<div class="sourceCode" id="cb109"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb109-1"><a href="module-3-generalized-linear-models.html#cb109-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(dat, <span class="fu">aes</span>(glucose, <span class="fu">factor</span>(diabetes)))</span>
<span id="cb109-2"><a href="module-3-generalized-linear-models.html#cb109-2" aria-hidden="true" tabindex="-1"></a>p <span class="sc">+</span> <span class="fu">geom_point</span>()</span></code></pre></div>
<pre><code>## Warning: Removed 5 rows containing missing values (`geom_point()`).</code></pre>
<p><img src="bookdown-demo_files/figure-html/plot-diabetes-vs-glucose-1.png" width="672" /></p>
<p>Could this be fit with a linear model?</p>
<div class="sourceCode" id="cb111"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb111-1"><a href="module-3-generalized-linear-models.html#cb111-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(dat, <span class="fu">aes</span>(glucose, <span class="fu">factor</span>(diabetes)))</span>
<span id="cb111-2"><a href="module-3-generalized-linear-models.html#cb111-2" aria-hidden="true" tabindex="-1"></a>p <span class="sc">+</span> <span class="fu">geom_point</span>() <span class="sc">+</span> <span class="fu">geom_smooth</span>()</span></code></pre></div>
<pre><code>## `geom_smooth()` using method = &#39;loess&#39; and formula = &#39;y ~ x&#39;</code></pre>
<pre><code>## Warning: Removed 5 rows containing non-finite values (`stat_smooth()`).</code></pre>
<pre><code>## Warning: Removed 5 rows containing missing values (`geom_point()`).</code></pre>
<p><img src="bookdown-demo_files/figure-html/unnamed-chunk-59-1.png" width="672" /></p>
<p>This is a situation where a logistic regression model would be an appropriate choice. We’re going to use <code>glm()</code> to fit a model to these data:</p>
<div class="sourceCode" id="cb115"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb115-1"><a href="module-3-generalized-linear-models.html#cb115-1" aria-hidden="true" tabindex="-1"></a>mod <span class="ot">&lt;-</span> <span class="fu">glm</span>(<span class="fu">factor</span>(diabetes)<span class="sc">~</span>glucose, </span>
<span id="cb115-2"><a href="module-3-generalized-linear-models.html#cb115-2" aria-hidden="true" tabindex="-1"></a>    dat,</span>
<span id="cb115-3"><a href="module-3-generalized-linear-models.html#cb115-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">family =</span> <span class="st">&quot;binomial&quot;</span> <span class="co"># set to model binary outcome</span></span>
<span id="cb115-4"><a href="module-3-generalized-linear-models.html#cb115-4" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb115-5"><a href="module-3-generalized-linear-models.html#cb115-5" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(mod)</span></code></pre></div>
<pre><code>## 
## Call:
## glm(formula = factor(diabetes) ~ glucose, family = &quot;binomial&quot;, 
##     data = dat)
## 
## Deviance Residuals: 
##     Min       1Q   Median       3Q      Max  
## -2.1846  -0.7773  -0.5094   0.8232   2.2896  
## 
## Coefficients:
##              Estimate Std. Error z value Pr(&gt;|z|)    
## (Intercept) -5.715088   0.438100  -13.04   &lt;2e-16 ***
## glucose      0.040634   0.003382   12.01   &lt;2e-16 ***
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## (Dispersion parameter for binomial family taken to be 1)
## 
##     Null deviance: 986.70  on 762  degrees of freedom
## Residual deviance: 786.56  on 761  degrees of freedom
##   (5 observations deleted due to missingness)
## AIC: 790.56
## 
## Number of Fisher Scoring iterations: 4</code></pre>
<p>Which factors explain a diabetes diagnosis?</p>
<p>What if we include a couple other factors?</p>
<div class="sourceCode" id="cb117"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb117-1"><a href="module-3-generalized-linear-models.html#cb117-1" aria-hidden="true" tabindex="-1"></a>mod <span class="ot">&lt;-</span> <span class="fu">glm</span>(<span class="fu">factor</span>(diabetes)<span class="sc">~</span> glucose <span class="sc">+</span> pregnant <span class="sc">+</span> age <span class="sc">+</span> pressure <span class="sc">+</span> triceps,</span>
<span id="cb117-2"><a href="module-3-generalized-linear-models.html#cb117-2" aria-hidden="true" tabindex="-1"></a>    dat,</span>
<span id="cb117-3"><a href="module-3-generalized-linear-models.html#cb117-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">family =</span> <span class="st">&quot;binomial&quot;</span>)</span>
<span id="cb117-4"><a href="module-3-generalized-linear-models.html#cb117-4" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(mod)</span></code></pre></div>
<pre><code>## 
## Call:
## glm(formula = factor(diabetes) ~ glucose + pregnant + age + pressure + 
##     triceps, family = &quot;binomial&quot;, data = dat)
## 
## Deviance Residuals: 
##     Min       1Q   Median       3Q      Max  
## -2.1790  -0.6813  -0.4061   0.7167   2.3204  
## 
## Coefficients:
##               Estimate Std. Error z value Pr(&gt;|z|)    
## (Intercept) -7.6685706  0.8451262  -9.074  &lt; 2e-16 ***
## glucose      0.0365926  0.0041105   8.902  &lt; 2e-16 ***
## pregnant     0.0972493  0.0417560   2.329 0.019860 *  
## age          0.0249707  0.0135945   1.837 0.066235 .  
## pressure    -0.0009321  0.0099322  -0.094 0.925230    
## triceps      0.0420045  0.0117044   3.589 0.000332 ***
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## (Dispersion parameter for binomial family taken to be 1)
## 
##     Null deviance: 678.40  on 533  degrees of freedom
## Residual deviance: 494.38  on 528  degrees of freedom
##   (234 observations deleted due to missingness)
## AIC: 506.38
## 
## Number of Fisher Scoring iterations: 5</code></pre>
<div class="rmd-caution">
<p>
Note: This morning’s session only intends to introduce you to fitting non-linear models.
</p>
<p>
In practice you may need to do more work to test multiple models to ascertain best fits your data, using measures such as goodness-of-fit. You will also likely compute odds ratio (odds of increased Y per unit increase X), which is out of scope for the current tutorial.
</p>
<p>
We strongly recommend that you learn these topics before applying these methods to your own data.
</p>
</div>
</div>
<div id="exercise-2" class="section level2 unnumbered">
<h2>Exercise</h2>
<p>Now let’s apply the ideas above to a dataset for classifying a breast cell as being either benign or malignant.</p>
<div class="sourceCode" id="cb119"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb119-1"><a href="module-3-generalized-linear-models.html#cb119-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(BreastCancer)</span>
<span id="cb119-2"><a href="module-3-generalized-linear-models.html#cb119-2" aria-hidden="true" tabindex="-1"></a>bc <span class="ot">&lt;-</span> BreastCancer</span>
<span id="cb119-3"><a href="module-3-generalized-linear-models.html#cb119-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (k <span class="cf">in</span> <span class="dv">2</span><span class="sc">:</span><span class="dv">10</span>) <span class="co"># altered for current lab</span></span>
<span id="cb119-4"><a href="module-3-generalized-linear-models.html#cb119-4" aria-hidden="true" tabindex="-1"></a>    bc[,k] <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(bc[,k]) </span>
<span id="cb119-5"><a href="module-3-generalized-linear-models.html#cb119-5" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(bc)</span></code></pre></div>
<pre><code>##        Id Cl.thickness Cell.size Cell.shape Marg.adhesion Epith.c.size Bare.nuclei
## 1 1000025            5         1          1             1            2           1
## 2 1002945            5         4          4             5            7          10
## 3 1015425            3         1          1             1            2           2
## 4 1016277            6         8          8             1            3           4
## 5 1017023            4         1          1             3            2           1
## 6 1017122            8        10         10             8            7          10
##   Bl.cromatin Normal.nucleoli Mitoses     Class
## 1           3               1       1    benign
## 2           3               2       1    benign
## 3           3               1       1    benign
## 4           3               7       1    benign
## 5           3               1       1    benign
## 6           9               7       1 malignant</code></pre>
<p>Learn more about the dataset:</p>
<div class="sourceCode" id="cb121"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb121-1"><a href="module-3-generalized-linear-models.html#cb121-1" aria-hidden="true" tabindex="-1"></a>?BreastCancer</span></code></pre></div>
<p>For your exercise, answer the following questions:</p>
<ul>
<li>Is there missing data?</li>
<li>Which columns are</li>
<li>Use plots to explore the relationship between explanatory variables.</li>
<li>Fit a logistic model to identify which factors explain cell size.</li>
</ul>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="module-2-exercise.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="module-3-exercise.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": null,
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["bookdown-demo.pdf", "bookdown-demo.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
