<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Module 4: Finding differentially expressed genes with RNA-seq | Analysis Using R</title>
  <meta name="description" content="Principles of exploratory data analysis, RNAseq differential expression, and generalized linear models" />
  <meta name="generator" content="bookdown 0.24 and GitBook 2.6.7" />

  <meta property="og:title" content="Module 4: Finding differentially expressed genes with RNA-seq | Analysis Using R" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Principles of exploratory data analysis, RNAseq differential expression, and generalized linear models" />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Module 4: Finding differentially expressed genes with RNA-seq | Analysis Using R" />
  
  <meta name="twitter:description" content="Principles of exploratory data analysis, RNAseq differential expression, and generalized linear models" />
  

<meta name="author" content="Instructors: Shraddha Pai, Chaitra Sarathy" />



  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  <link rel="shortcut icon" href="images/favicon.ico" type="image/x-icon" />
<link rel="prev" href="module-3-exercise.html"/>
<link rel="next" href="module-4---exercises.html"/>
<script src="libs/header-attrs-2.11/header-attrs.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<img src="images/bioinfologo_big.png">
<li><a href="./">2023 CBW Analysis Using R Workshop</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#meet-your-faculty"><i class="fa fa-check"></i>Meet your Faculty</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#pre-workshop-materials-and-laptop-setup-instructions"><i class="fa fa-check"></i>Pre-workshop Materials and Laptop Setup Instructions</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#laptop-setup-instructions"><i class="fa fa-check"></i>Laptop Setup Instructions</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#r-packages"><i class="fa fa-check"></i>R packages</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="lecture-slides.html"><a href="lecture-slides.html"><i class="fa fa-check"></i>Lecture slides</a></li>
<li class="chapter" data-level="" data-path="module-1-exploratory-data-analysis-and-clustering.html"><a href="module-1-exploratory-data-analysis-and-clustering.html"><i class="fa fa-check"></i>Module 1: Exploratory Data Analysis and Clustering</a>
<ul>
<li><a href="module-1-exploratory-data-analysis-and-clustering.html#load-mouse-data">Load <code>mouse</code> data</a></li>
<li><a href="module-1-exploratory-data-analysis-and-clustering.html#rcolorbrewer-for-colour-palettes"><code>RColorBrewer</code> for colour palettes</a></li>
<li class="chapter" data-level="" data-path="module-1-exploratory-data-analysis-and-clustering.html"><a href="module-1-exploratory-data-analysis-and-clustering.html#correlations-distances-and-clustering"><i class="fa fa-check"></i>Correlations, distances, and clustering</a></li>
<li class="chapter" data-level="" data-path="module-1-exploratory-data-analysis-and-clustering.html"><a href="module-1-exploratory-data-analysis-and-clustering.html#hierarchical-clustering"><i class="fa fa-check"></i>Hierarchical clustering</a></li>
<li class="chapter" data-level="" data-path="module-1-exploratory-data-analysis-and-clustering.html"><a href="module-1-exploratory-data-analysis-and-clustering.html#k-means-clustering"><i class="fa fa-check"></i>K-means clustering</a></li>
<li><a href="module-1-exploratory-data-analysis-and-clustering.html#using-clvalid-to-determine-number-of-clusters">Using <code>clValid</code> to determine number of clusters</a></li>
<li class="chapter" data-level="" data-path="module-1-exploratory-data-analysis-and-clustering.html"><a href="module-1-exploratory-data-analysis-and-clustering.html#exercise"><i class="fa fa-check"></i>Exercise</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-1-exercise-results.html"><a href="module-1-exercise-results.html"><i class="fa fa-check"></i>Module 1 Exercise Results</a></li>
<li class="chapter" data-level="" data-path="module-2-dimensionality-reduction.html"><a href="module-2-dimensionality-reduction.html"><i class="fa fa-check"></i>Module 2: Dimensionality reduction</a>
<ul>
<li class="chapter" data-level="" data-path="module-2-dimensionality-reduction.html"><a href="module-2-dimensionality-reduction.html#exercise-1"><i class="fa fa-check"></i>Exercise</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-2-exercise.html"><a href="module-2-exercise.html"><i class="fa fa-check"></i>Module 2 exercise</a></li>
<li class="chapter" data-level="" data-path="module-3-generalized-linear-models.html"><a href="module-3-generalized-linear-models.html"><i class="fa fa-check"></i>Module 3: Generalized Linear Models</a>
<ul>
<li class="chapter" data-level="" data-path="module-3-generalized-linear-models.html"><a href="module-3-generalized-linear-models.html#read-data-from-files-and-explore-variable-distribution"><i class="fa fa-check"></i>Read data from files and explore variable distribution</a></li>
<li class="chapter" data-level="" data-path="module-3-generalized-linear-models.html"><a href="module-3-generalized-linear-models.html#explore-missing-data"><i class="fa fa-check"></i>Explore missing data</a></li>
<li><a href="module-3-generalized-linear-models.html#create-plots-with-ggplot2-to-explore-variable-relationships">Create plots with <code>ggplot2</code> to explore variable relationships</a></li>
<li><a href="module-3-generalized-linear-models.html#fit-binary-response-variable-using-glm-and-logistic-regression">Fit binary response variable using <code>glm()</code> and logistic regression</a></li>
<li class="chapter" data-level="" data-path="module-3-generalized-linear-models.html"><a href="module-3-generalized-linear-models.html#exercise-2"><i class="fa fa-check"></i>Exercise</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-3-exercise.html"><a href="module-3-exercise.html"><i class="fa fa-check"></i>Module 3 exercise</a></li>
<li class="chapter" data-level="" data-path="module-4-finding-differentially-expressed-genes-with-rna-seq.html"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html"><i class="fa fa-check"></i>Module 4: Finding differentially expressed genes with RNA-seq</a>
<ul>
<li class="chapter" data-level="" data-path="module-4-finding-differentially-expressed-genes-with-rna-seq.html"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#mini-introduction-to-bioconductor"><i class="fa fa-check"></i>Mini introduction to BioConductor</a></li>
<li><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#fetch-breast-cancer-data-using-curatedtcgadata">Fetch breast cancer data using <code>curatedTCGAData</code></a>
<ul>
<li class="chapter" data-level="" data-path="module-4-finding-differentially-expressed-genes-with-rna-seq.html"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#prepare-data-for-differential-expression-analysis"><i class="fa fa-check"></i>Prepare data for differential expression analysis</a></li>
</ul></li>
<li><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#differential-expression-analysis-with-edger">Differential expression analysis with <code>edgeR</code></a></li>
<li class="chapter" data-level="" data-path="module-4-finding-differentially-expressed-genes-with-rna-seq.html"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#exercise-3"><i class="fa fa-check"></i>Exercise</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-4---exercises.html"><a href="module-4---exercises.html"><i class="fa fa-check"></i>Module 4 - exercises</a></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Analysis Using R</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="module-4-finding-differentially-expressed-genes-with-rna-seq" class="section level1 unnumbered">
<h1>Module 4: Finding differentially expressed genes with RNA-seq</h1>
<p>We are going to look at genes differentially expressed between Luminal and Basal subtypes of breast cancer.</p>
<p>For this we are going to download data from The Cancer Genome Atlas [ADD LINK] readily available in BioConductor.</p>
<div id="mini-introduction-to-bioconductor" class="section level2 unnumbered">
<h2>Mini introduction to BioConductor</h2>
<p>BioConductor is the most popular software repository for genomics data analysis, and is worth exploring in detail.
This R-based ecosystem is open-source and community built. It contains software <strong>LINK</strong> packages but also <strong>DATA</strong> and <strong>ANNOTATION</strong> packages.</p>
<p>Examples of packages in BioConductor:</p>
<ul>
<li>Data package: The entire human genome sequence e.g., for the current GRCH38 build</li>
<li>Data package: Multi-modal genomic data from The Cancer Genome Atlas for various tumour types</li>
<li>Annotation package: Annotation for the Illumina DNA methylation array</li>
<li>Software: <code>DEseq</code> to process RNAseq data, <code>minfi</code> to process DNA methylation array data, CHIPseeker for Chip-Seq data, etc.,</li>
</ul>
<p>As we will see, BioConductor also has specialized <em>data structures</em> for bioinformatics operations.
Data structures are specialized, standardized formats for organizing, accessing, modifying and storing data.</p>
<p>Examples of BioConductor data structures:</p>
<ul>
<li><code>GenomicRanges</code>: for working with genomic coordinates</li>
<li><code>SummarizedExperiment</code>: container for storing data and metadata about an experiment</li>
<li><code>MultiAssayExperiment</code>: container for experiments where multiple genomic assays were run on the same samples</li>
</ul>
<p>BioConductor uses its own package manager to install packages. Instead of <code>install.packages()</code> we use, <code>BiocManager::install()</code>.</p>
<div class="sourceCode" id="cb124"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb124-1"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb124-1" aria-hidden="true" tabindex="-1"></a> <span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">&quot;curatedTCGAData&quot;</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>)) </span>
<span id="cb124-2"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb124-2" aria-hidden="true" tabindex="-1"></a>    BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">&quot;curatedTCGAData&quot;</span>)</span></code></pre></div>
</div>
<div id="fetch-breast-cancer-data-using-curatedtcgadata" class="section level2 unnumbered">
<h2>Fetch breast cancer data using <code>curatedTCGAData</code></h2>
<p>Load the package:</p>
<div class="sourceCode" id="cb125"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb125-1"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb125-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(curatedTCGAData)</span></code></pre></div>
<p>Let’s take a look at the available data for breast cancer, without downloading anything (set <code>dry.run=TRUE</code>).
Each row shows a data layer available. e.g., “<em>mRNA</em>” is transcriptomic data, “<em>CNA</em>” is chromosomal copy number aberration data, etc.,</p>
<div class="sourceCode" id="cb126"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb126-1"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb126-1" aria-hidden="true" tabindex="-1"></a><span class="fu">curatedTCGAData</span>(<span class="at">diseaseCode=</span><span class="st">&quot;BRCA&quot;</span>, <span class="at">assays=</span><span class="st">&quot;*&quot;</span>,<span class="at">dry.run=</span><span class="cn">TRUE</span>, <span class="at">version=</span><span class="st">&quot;2.0.1&quot;</span>)</span></code></pre></div>
<pre><code>## snapshotDate(): 2021-10-19</code></pre>
<pre><code>## See &#39;?curatedTCGAData&#39; for &#39;diseaseCode&#39; and &#39;assays&#39; inputs</code></pre>
<pre><code>##     ah_id                                      title file_size
## 1  EH4769                       BRCA_CNASeq-20160128      0 Mb
## 2  EH4770                       BRCA_CNASNP-20160128    9.8 Mb
## 3  EH4771                       BRCA_CNVSNP-20160128    2.8 Mb
## 4  EH4773             BRCA_GISTIC_AllByGene-20160128    1.2 Mb
## 5  EH4774                 BRCA_GISTIC_Peaks-20160128      0 Mb
## 6  EH4775     BRCA_GISTIC_ThresholdedByGene-20160128    0.3 Mb
## 7  EH4777  BRCA_Methylation_methyl27-20160128_assays   60.7 Mb
## 8  EH4778      BRCA_Methylation_methyl27-20160128_se    0.4 Mb
## 9  EH4779 BRCA_Methylation_methyl450-20160128_assays 2646.4 Mb
## 10 EH4780     BRCA_Methylation_methyl450-20160128_se      6 Mb
## 11 EH4781                 BRCA_miRNASeqGene-20160128    0.6 Mb
## 12 EH4782                    BRCA_mRNAArray-20160128   27.3 Mb
## 13 EH4783                     BRCA_Mutation-20160128    4.5 Mb
## 14 EH4784                  BRCA_RNASeq2Gene-20160128   43.1 Mb
## 15 EH4785              BRCA_RNASeq2GeneNorm-20160128   64.5 Mb
## 16 EH4786                   BRCA_RNASeqGene-20160128     30 Mb
## 17 EH4787                    BRCA_RPPAArray-20160128    1.5 Mb
##                    rdataclass rdatadateadded rdatadateremoved
## 1            RaggedExperiment     2021-01-27             &lt;NA&gt;
## 2            RaggedExperiment     2021-01-27             &lt;NA&gt;
## 3            RaggedExperiment     2021-01-27             &lt;NA&gt;
## 4        SummarizedExperiment     2021-01-27             &lt;NA&gt;
## 5  RangedSummarizedExperiment     2021-01-27             &lt;NA&gt;
## 6        SummarizedExperiment     2021-01-27             &lt;NA&gt;
## 7        SummarizedExperiment     2021-01-27             &lt;NA&gt;
## 8        SummarizedExperiment     2021-01-27             &lt;NA&gt;
## 9            RaggedExperiment     2021-01-27             &lt;NA&gt;
## 10       SummarizedExperiment     2021-01-27             &lt;NA&gt;
## 11       SummarizedExperiment     2021-01-27             &lt;NA&gt;
## 12       SummarizedExperiment     2021-01-27             &lt;NA&gt;
## 13       SummarizedExperiment     2021-01-27             &lt;NA&gt;
## 14                     DFrame     2021-01-27             &lt;NA&gt;
## 15       SummarizedExperiment     2021-01-27             &lt;NA&gt;
## 16       SummarizedExperiment     2021-01-27             &lt;NA&gt;
## 17       SummarizedExperiment     2021-01-27             &lt;NA&gt;</code></pre>
<p>We want to get unprocessed gene expression read counts, so let’s fetch the <code>BRCA_RNASeq2Gene-20160128</code> layer. You can use <code>?curatedTCGAData</code> to see a description of all data layers.</p>
<div class="sourceCode" id="cb130"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb130-1"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb130-1" aria-hidden="true" tabindex="-1"></a>brca <span class="ot">&lt;-</span> <span class="fu">suppressMessages</span>(</span>
<span id="cb130-2"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb130-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">curatedTCGAData</span>(</span>
<span id="cb130-3"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb130-3" aria-hidden="true" tabindex="-1"></a>        <span class="st">&quot;BRCA&quot;</span>,</span>
<span id="cb130-4"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb130-4" aria-hidden="true" tabindex="-1"></a>        <span class="at">assays=</span><span class="st">&quot;RNASeqGene&quot;</span>,</span>
<span id="cb130-5"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb130-5" aria-hidden="true" tabindex="-1"></a>        <span class="at">dry.run=</span><span class="cn">FALSE</span>, </span>
<span id="cb130-6"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb130-6" aria-hidden="true" tabindex="-1"></a>        <span class="at">version=</span><span class="st">&quot;2.0.1&quot;</span>)</span>
<span id="cb130-7"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb130-7" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p>This call returns a <code>MultiAssayExperiment</code> object. Recall that this is a container for storing multiple assays performed on the same set of samples. <a href="https://bioconductor.org/packages/release/bioc/vignettes/MultiAssayExperiment/inst/doc/QuickStartMultiAssay.html">See this tutorial</a> to learn more.</p>
<p>Let’s briefly explore the <code>brca</code> <code>MultiAssayExperiment</code> object.</p>
<div class="sourceCode" id="cb131"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb131-1"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb131-1" aria-hidden="true" tabindex="-1"></a>brca</span></code></pre></div>
<pre><code>## A MultiAssayExperiment object of 1 listed
##  experiment with a user-defined name and respective class.
##  Containing an ExperimentList class object of length 1:
##  [1] BRCA_RNASeqGene-20160128: SummarizedExperiment with 20502 rows and 878 columns
## Functionality:
##  experiments() - obtain the ExperimentList instance
##  colData() - the primary/phenotype DataFrame
##  sampleMap() - the sample coordination DataFrame
##  `$`, `[`, `[[` - extract colData columns, subset, or experiment
##  *Format() - convert into a long or wide DataFrame
##  assays() - convert ExperimentList to a SimpleList of matrices
##  exportClass() - save data to flat files</code></pre>
<p><code>assays()</code> returns a <code>list</code> with all -omic data associated with this object. Here we just have the one we downloaded.</p>
<div class="sourceCode" id="cb133"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb133-1"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb133-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(<span class="fu">assays</span>(brca))</span></code></pre></div>
<pre><code>## [1] &quot;List object of length 1 with 0 metadata columns&quot;</code></pre>
<p><code>names()</code> shows the datatypes in each slot of <code>assays()</code>:</p>
<div class="sourceCode" id="cb135"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb135-1"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb135-1" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(<span class="fu">assays</span>(brca))</span></code></pre></div>
<pre><code>## [1] &quot;BRCA_RNASeqGene-20160128&quot;</code></pre>
<p>So gene expression in slot 1.</p>
<p>We can subset the data to see what it looks like. Let’s look at just the first five measures, for the first 10 samples:</p>
<div class="sourceCode" id="cb137"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb137-1"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb137-1" aria-hidden="true" tabindex="-1"></a>xpr <span class="ot">&lt;-</span> <span class="fu">assays</span>(brca)[[<span class="dv">1</span>]]</span>
<span id="cb137-2"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb137-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(xpr[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>])</span></code></pre></div>
<pre><code>##       TCGA-A1-A0SB-01A-11R-A144-07 TCGA-A1-A0SD-01A-11R-A115-07
## A1BG                           164                          546
## A1CF                             0                            0
## A2BP1                           22                            1
## A2LD1                          127                          331
## A2ML1                           94                          144
## A2M                         102123                       107181
##       TCGA-A1-A0SE-01A-11R-A084-07 TCGA-A1-A0SF-01A-11R-A144-07
## A1BG                          1341                          836
## A1CF                             0                            1
## A2BP1                            2                            0
## A2LD1                          498                          526
## A2ML1                          114                           77
## A2M                         101192                        50316
##       TCGA-A1-A0SG-01A-11R-A144-07
## A1BG                           512
## A1CF                             3
## A2BP1                           25
## A2LD1                          451
## A2ML1                           76
## A2M                          45826</code></pre>
<p>How many measures do we have?</p>
<div class="sourceCode" id="cb139"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb139-1"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb139-1" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>(xpr)</span></code></pre></div>
<pre><code>## [1] 20502</code></pre>
<div id="prepare-data-for-differential-expression-analysis" class="section level3 unnumbered">
<h3>Prepare data for differential expression analysis</h3>
<p>Process the expression values:</p>
<div class="sourceCode" id="cb141"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb141-1"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb141-1" aria-hidden="true" tabindex="-1"></a>cnames <span class="ot">&lt;-</span> <span class="fu">colnames</span>(xpr)</span>
<span id="cb141-2"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb141-2" aria-hidden="true" tabindex="-1"></a>hpos <span class="ot">&lt;-</span> <span class="fu">gregexpr</span>(<span class="st">&quot;-&quot;</span>,cnames)</span>
<span id="cb141-3"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb141-3" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">&lt;-</span> <span class="fu">sapply</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(cnames),<span class="cf">function</span>(i) {<span class="fu">substr</span>(cnames[i],<span class="dv">1</span>,hpos[[i]][<span class="dv">3</span>]<span class="sc">-</span><span class="dv">1</span>)})</span>
<span id="cb141-4"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb141-4" aria-hidden="true" tabindex="-1"></a>idx <span class="ot">&lt;-</span> <span class="sc">!</span><span class="fu">duplicated</span>(tmp)</span>
<span id="cb141-5"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb141-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb141-6"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb141-6" aria-hidden="true" tabindex="-1"></a><span class="co"># subset values</span></span>
<span id="cb141-7"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb141-7" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">&lt;-</span> tmp[idx]</span>
<span id="cb141-8"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb141-8" aria-hidden="true" tabindex="-1"></a>xpr <span class="ot">&lt;-</span> xpr[,idx]</span>
<span id="cb141-9"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb141-9" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(xpr) <span class="ot">&lt;-</span> tmp</span></code></pre></div>
<p>Patient metadata is contained in the <code>colData()</code> slot. Rows contain data for each patient and columns contain measures such as clinical characteristics:</p>
<div class="sourceCode" id="cb142"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb142-1"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb142-1" aria-hidden="true" tabindex="-1"></a>pheno <span class="ot">&lt;-</span> <span class="fu">colData</span>(brca)</span>
<span id="cb142-2"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb142-2" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(pheno)[<span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>]</span></code></pre></div>
<pre><code>##  [1] &quot;patientID&quot;                            &quot;years_to_birth&quot;                      
##  [3] &quot;vital_status&quot;                         &quot;days_to_death&quot;                       
##  [5] &quot;days_to_last_followup&quot;                &quot;tumor_tissue_site&quot;                   
##  [7] &quot;pathologic_stage&quot;                     &quot;pathology_T_stage&quot;                   
##  [9] &quot;pathology_N_stage&quot;                    &quot;pathology_M_stage&quot;                   
## [11] &quot;gender&quot;                               &quot;date_of_initial_pathologic_diagnosis&quot;
## [13] &quot;days_to_last_known_alive&quot;             &quot;radiation_therapy&quot;                   
## [15] &quot;histological_type&quot;                    &quot;number_of_lymph_nodes&quot;               
## [17] &quot;race&quot;                                 &quot;ethnicity&quot;                           
## [19] &quot;admin.bcr&quot;                            &quot;admin.day_of_dcc_upload&quot;</code></pre>
<div class="sourceCode" id="cb144"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb144-1"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb144-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(pheno[,<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>])</span></code></pre></div>
<pre><code>## DataFrame with 6 rows and 5 columns
##                 patientID years_to_birth vital_status days_to_death
##               &lt;character&gt;      &lt;integer&gt;    &lt;integer&gt;     &lt;integer&gt;
## TCGA-A1-A0SB TCGA-A1-A0SB             70            0            NA
## TCGA-A1-A0SD TCGA-A1-A0SD             59            0            NA
## TCGA-A1-A0SE TCGA-A1-A0SE             56            0            NA
## TCGA-A1-A0SF TCGA-A1-A0SF             54            0            NA
## TCGA-A1-A0SG TCGA-A1-A0SG             61            0            NA
## TCGA-A1-A0SH TCGA-A1-A0SH             39            0            NA
##              days_to_last_followup
##                          &lt;integer&gt;
## TCGA-A1-A0SB                   259
## TCGA-A1-A0SD                   437
## TCGA-A1-A0SE                  1321
## TCGA-A1-A0SF                  1463
## TCGA-A1-A0SG                   434
## TCGA-A1-A0SH                  1437</code></pre>
<p>Let’s confirm that samples are in the same order in the <code>xpr</code> matrix and in the metadata table <code>pheno</code>.</p>
<div class="sourceCode" id="cb146"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb146-1"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb146-1" aria-hidden="true" tabindex="-1"></a>ids_in_both <span class="ot">&lt;-</span> <span class="fu">intersect</span>(pheno<span class="sc">$</span>patientID, <span class="fu">colnames</span>(xpr))</span>
<span id="cb146-2"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb146-2" aria-hidden="true" tabindex="-1"></a>pheno <span class="ot">&lt;-</span> pheno[<span class="fu">which</span>(pheno<span class="sc">$</span>patientID <span class="sc">%in%</span> ids_in_both),]</span>
<span id="cb146-3"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb146-3" aria-hidden="true" tabindex="-1"></a>xpr <span class="ot">&lt;-</span> xpr[,<span class="fu">which</span>(<span class="fu">colnames</span>(xpr) <span class="sc">%in%</span> ids_in_both)]</span>
<span id="cb146-4"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb146-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb146-5"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb146-5" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">all.equal</span>(pheno<span class="sc">$</span>patientID, <span class="fu">colnames</span>(xpr))<span class="sc">!=</span><span class="cn">TRUE</span>) {</span>
<span id="cb146-6"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb146-6" aria-hidden="true" tabindex="-1"></a>    midx <span class="ot">&lt;-</span> <span class="fu">match</span>(pheno<span class="sc">$</span>patientID, <span class="fu">colnames</span>(xpr))</span>
<span id="cb146-7"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb146-7" aria-hidden="true" tabindex="-1"></a>    xpr <span class="ot">&lt;-</span> xpr[,midx]</span>
<span id="cb146-8"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb146-8" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<div class="rmd-caution">
<p>
Always check that samples are in the same order in the tables you are comparing, and use <code>match()</code> to reorder them if necessary. If they are not in the same order, you are matching data to the wrong sample, and your results will be wrong.
</p>
</div>
<p>This table has many columns. Let’s just keep those related to tumour type according to the PAM50 classification.
What tumour types do we have in this dataset?</p>
<div class="sourceCode" id="cb147"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb147-1"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb147-1" aria-hidden="true" tabindex="-1"></a>pheno <span class="ot">&lt;-</span> pheno[,<span class="fu">c</span>(<span class="st">&quot;patientID&quot;</span>,<span class="st">&quot;PAM50.mRNA&quot;</span>)]</span>
<span id="cb147-2"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb147-2" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(pheno<span class="sc">$</span>PAM50.mRNA)</span></code></pre></div>
<pre><code>## 
##    Basal-like HER2-enriched     Luminal A     Luminal B   Normal-like 
##            90            55           216           117             7</code></pre>
<p>Let’s limit the comparison to <code>Luminal A</code> and <code>Basal</code> type tumours.</p>
<div class="sourceCode" id="cb149"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb149-1"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb149-1" aria-hidden="true" tabindex="-1"></a>idx <span class="ot">&lt;-</span> <span class="fu">which</span>(pheno<span class="sc">$</span>PAM50.mRNA <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">&quot;Luminal A&quot;</span>,<span class="st">&quot;Basal-like&quot;</span>))</span>
<span id="cb149-2"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb149-2" aria-hidden="true" tabindex="-1"></a>pheno <span class="ot">&lt;-</span> pheno[idx,]</span>
<span id="cb149-3"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb149-3" aria-hidden="true" tabindex="-1"></a>xpr <span class="ot">&lt;-</span> xpr[,idx] </span>
<span id="cb149-4"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb149-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb149-5"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb149-5" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(pheno)</span></code></pre></div>
<pre><code>## [1] 306   2</code></pre>
<div class="sourceCode" id="cb151"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb151-1"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb151-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(xpr)</span></code></pre></div>
<pre><code>## [1] 20502   306</code></pre>
</div>
</div>
<div id="differential-expression-analysis-with-edger" class="section level2 unnumbered">
<h2>Differential expression analysis with <code>edgeR</code></h2>
<p>Now that the data are prepared, let’s created a <code>DGEList</code> object (DGE stands for “Differential Gene Expression”). This object is what we will use for our differential expression analysis.</p>
<p>Note: Make phenotype of interest categorical. In R that means converting to a <code>factor</code> type with categorical <code>levels</code>. You can think of levels as ordinal representations (e.g., first level = 1, second = 2, etc., )</p>
<p>If <code>levels=</code> are not set, the default uses alphabetical order. We recommend explicitly setting levels so that there are no assumptions.</p>
<p>Load the <code>edgeR</code> package:</p>
<div class="sourceCode" id="cb153"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb153-1"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb153-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">require</span>(edgeR))</span></code></pre></div>
<p>Let’s create a <code>DGEList</code> object for the differential expression analysis. Note that <code>group</code> must be a categorical variable (use <code>factor()</code> to convert it to one):</p>
<div class="sourceCode" id="cb154"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb154-1"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb154-1" aria-hidden="true" tabindex="-1"></a>group <span class="ot">&lt;-</span> <span class="fu">factor</span>(</span>
<span id="cb154-2"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb154-2" aria-hidden="true" tabindex="-1"></a>        pheno<span class="sc">$</span>PAM50.mRNA,</span>
<span id="cb154-3"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb154-3" aria-hidden="true" tabindex="-1"></a>        <span class="at">levels=</span><span class="fu">c</span>(<span class="st">&quot;Luminal A&quot;</span>,<span class="st">&quot;Basal-like&quot;</span>)</span>
<span id="cb154-4"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb154-4" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb154-5"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb154-5" aria-hidden="true" tabindex="-1"></a>dge <span class="ot">&lt;-</span> <span class="fu">DGEList</span>(</span>
<span id="cb154-6"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb154-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">counts =</span> xpr,</span>
<span id="cb154-7"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb154-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">group =</span> group</span>
<span id="cb154-8"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb154-8" aria-hidden="true" tabindex="-1"></a>    )</span></code></pre></div>
<p>Remove low-count genes: Only keep genes with 100 or greater counts per million for at least two samples:</p>
<div class="sourceCode" id="cb155"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb155-1"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb155-1" aria-hidden="true" tabindex="-1"></a>(dge<span class="sc">$</span>counts[<span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>])</span></code></pre></div>
<pre><code>##       TCGA-A1-A0SD TCGA-A1-A0SE TCGA-A1-A0SH TCGA-A1-A0SJ TCGA-A1-A0SK TCGA-A1-A0SO
## A1BG           546         1341         1126         1433          626         2502
## A1CF             0            0            1            0            1            4
## A2BP1            1            2            4            1            1            0
## A2LD1          331          498          838          831         2948          883
## A2ML1          144          114          127          119          367          335
## A2M         107181       101192        81857       128591        27972        32286
##       TCGA-A2-A04N TCGA-A2-A04P TCGA-A2-A04Q TCGA-A2-A04T TCGA-A2-A04U TCGA-A2-A04V
## A1BG           592          503          660          485         1269          870
## A1CF             1            4            1            3            5            2
## A2BP1            0            2            1           11            0            7
## A2LD1          487          384          694          534          200          454
## A2ML1           87         7435        13543        89424         3088          160
## A2M          35748        38122        70012        76814       178113        50336
##       TCGA-A2-A04Y TCGA-A2-A0CM TCGA-A2-A0CP TCGA-A2-A0CQ TCGA-A2-A0CS TCGA-A2-A0CU
## A1BG          1392          297         1240          919         1422          688
## A1CF             1            0            0            1            0            0
## A2BP1           18            0            2            1            8            1
## A2LD1          932          161          335          111          589          329
## A2ML1          250         4770           76          217           65          168
## A2M          75809        37519        76407        45968        18353       197374
##       TCGA-A2-A0CV TCGA-A2-A0CZ
## A1BG           532         1628
## A1CF             0            4
## A2BP1            6            3
## A2LD1          458          468
## A2ML1          162          102
## A2M          85242        85978</code></pre>
<p>Look at counts per million using <code>cpm</code>:</p>
<div class="sourceCode" id="cb157"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb157-1"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb157-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cpm</span>(dge)[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code></pre></div>
<pre><code>##       TCGA-A1-A0SD TCGA-A1-A0SE TCGA-A1-A0SH TCGA-A1-A0SJ TCGA-A1-A0SK
## A1BG    6.40538988  10.40014782  7.874879329 11.979028558  4.544931080
## A1CF    0.00000000   0.00000000  0.006993676  0.000000000  0.007260273
## A2BP1   0.01173148   0.01551103  0.027974705  0.008359406  0.007260273
## A2LD1   3.88312097   3.86224729  5.860700602  6.946666247 21.403285660
## A2ML1   1.68933359   0.88412890  0.888196869  0.994769294  2.664520298</code></pre>
<p>This next line is a bit complex so let’s unpack it:</p>
<ul>
<li>We are using <code>cpm(dge)&gt;100</code> as a logical test (“which genes have cpm &gt; 100?”).</li>
<li>For each gene, we want that test to be true for at least two samples. For this we use <code>rowSums()</code> to add up how many samples meet that criteria.</li>
</ul>
<div class="sourceCode" id="cb159"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb159-1"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb159-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(dge) <span class="co">#before </span></span></code></pre></div>
<pre><code>## [1] 20502   306</code></pre>
<div class="sourceCode" id="cb161"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb161-1"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb161-1" aria-hidden="true" tabindex="-1"></a>tokeep <span class="ot">&lt;-</span> <span class="fu">rowSums</span>(<span class="fu">cpm</span>(dge)<span class="sc">&gt;</span><span class="dv">100</span>) <span class="sc">&gt;=</span> <span class="dv">2</span></span>
<span id="cb161-2"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb161-2" aria-hidden="true" tabindex="-1"></a>dge <span class="ot">&lt;-</span> dge[tokeep,,keep.lib.sizes <span class="ot">=</span> <span class="cn">FALSE</span>]</span>
<span id="cb161-3"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb161-3" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(dge) <span class="co">#after</span></span></code></pre></div>
<pre><code>## [1] 7699  306</code></pre>
<p>Normalize the data:</p>
<div class="sourceCode" id="cb163"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb163-1"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb163-1" aria-hidden="true" tabindex="-1"></a>dge <span class="ot">&lt;-</span> <span class="fu">calcNormFactors</span>(dge)</span></code></pre></div>
<p>Visualize the data:</p>
<div class="sourceCode" id="cb164"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb164-1"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb164-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotMDS</span>(</span>
<span id="cb164-2"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb164-2" aria-hidden="true" tabindex="-1"></a>    dge, </span>
<span id="cb164-3"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb164-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">col=</span><span class="fu">as.numeric</span>(dge<span class="sc">$</span>samples<span class="sc">$</span>group), </span>
<span id="cb164-4"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb164-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">pch=</span><span class="dv">16</span></span>
<span id="cb164-5"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb164-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb164-6"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb164-6" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(</span>
<span id="cb164-7"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb164-7" aria-hidden="true" tabindex="-1"></a>    <span class="st">&quot;bottomleft&quot;</span>, </span>
<span id="cb164-8"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb164-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as.character</span>(<span class="fu">unique</span>(dge<span class="sc">$</span>samples<span class="sc">$</span>group)),</span>
<span id="cb164-9"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb164-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">col=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>), <span class="at">pch=</span><span class="dv">16</span></span>
<span id="cb164-10"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb164-10" aria-hidden="true" tabindex="-1"></a>    )</span></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/unnamed-chunk-70-1.png" width="672" /></p>
<p>Let’s create a model design to identify genes with a <code>group</code> effect:</p>
<div class="sourceCode" id="cb165"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb165-1"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb165-1" aria-hidden="true" tabindex="-1"></a>group <span class="ot">&lt;-</span> dge<span class="sc">$</span>samples<span class="sc">$</span>group</span>
<span id="cb165-2"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb165-2" aria-hidden="true" tabindex="-1"></a>mod <span class="ot">&lt;-</span> <span class="fu">model.matrix</span>(<span class="sc">~</span>group)</span></code></pre></div>
<p>Estimate variation (“dispersion”) for each gene:</p>
<div class="sourceCode" id="cb166"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb166-1"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb166-1" aria-hidden="true" tabindex="-1"></a>dge <span class="ot">&lt;-</span> <span class="fu">estimateDisp</span>(dge, mod)</span></code></pre></div>
<p>Call differentially expressed genes.</p>
<p>Here we:</p>
<ul>
<li>fit a model for each gene, using <code>glmFit</code></li>
<li>we have built in an estimate of gene-wise dispersion to better identify treatment effect (or “contrast”)</li>
<li>for each gene, we run a likelihood ratio test which compares which model fits the data better: a null model (treatment effect = 0)
or a full model (treatment effect is non-zero)</li>
</ul>
<p>Note that <code>coef=2</code> fetches the effects for the treatment effect; <code>coef=1</code> would fetch effects of the intercept term.</p>
<div class="sourceCode" id="cb167"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb167-1"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb167-1" aria-hidden="true" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">glmFit</span>(dge,mod)</span>
<span id="cb167-2"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb167-2" aria-hidden="true" tabindex="-1"></a>diffEx <span class="ot">&lt;-</span> <span class="fu">glmLRT</span>(fit, <span class="at">coef =</span> <span class="dv">2</span>) <span class="co"># get coefficients for group term</span></span></code></pre></div>
<p>Look at the top 10 differentially expressed genes:</p>
<div class="sourceCode" id="cb168"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb168-1"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb168-1" aria-hidden="true" tabindex="-1"></a>tt <span class="ot">&lt;-</span> <span class="fu">topTags</span>(diffEx, <span class="at">n=</span><span class="dv">10</span>)</span>
<span id="cb168-2"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb168-2" aria-hidden="true" tabindex="-1"></a>tt</span></code></pre></div>
<pre><code>## Coefficient:  groupBasal-like 
##             logFC   logCPM       LR        PValue           FDR
## ART3     4.437401 3.455108 795.7399 4.552774e-175 3.505180e-171
## EN1      4.493149 4.469186 568.8034 1.022382e-125 3.935658e-122
## MTHFD1L  2.072720 4.744988 548.6874 2.429549e-121 6.235032e-118
## CLDN6    5.917771 1.735389 528.0710 7.423651e-117 1.428867e-113
## CENPW    2.628365 3.806891 523.0933 8.985654e-116 1.383611e-112
## IL12RB2  3.981078 3.080989 507.2309 2.539169e-112 3.258177e-109
## TTLL4    1.857331 5.441088 504.0036 1.279029e-111 1.406749e-108
## CXorf49B 6.007052 1.678586 484.6339 2.096210e-107 2.017340e-104
## GFRA3    4.890609 1.725042 468.7621 5.958600e-104 5.097251e-101
## HAPLN3   2.737617 5.215775 460.3033 4.129466e-102  3.179276e-99</code></pre>
<p>For the next steps we’re going to need stats on all the genes we’ve tested. So let’s get those:</p>
<div class="sourceCode" id="cb170"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb170-1"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb170-1" aria-hidden="true" tabindex="-1"></a>tt <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(</span>
<span id="cb170-2"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb170-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">topTags</span>(diffEx, <span class="at">n=</span><span class="fu">nrow</span>(dge)</span>
<span id="cb170-3"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb170-3" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb170-4"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb170-4" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p>P-value histogram gives us a sense of the false positive rate.</p>
<p>For data without any signal we may expect a p-value histogram that is randomly uniformly distributed between 0 and 1. By chance, 5% of the time we expect p-values &lt; 0.05.</p>
<div class="sourceCode" id="cb171"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb171-1"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb171-1" aria-hidden="true" tabindex="-1"></a>rnd_pval <span class="ot">&lt;-</span> <span class="fu">runif</span>(<span class="fu">nrow</span>(tt)) </span>
<span id="cb171-2"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb171-2" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(</span>
<span id="cb171-3"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb171-3" aria-hidden="true" tabindex="-1"></a>    rnd_pval,</span>
<span id="cb171-4"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb171-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">n=</span><span class="dv">100</span>,</span>
<span id="cb171-5"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb171-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">main=</span><span class="st">&quot;Histogram for uniformly-distributed data&quot;</span></span>
<span id="cb171-6"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb171-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb171-7"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb171-7" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">v=</span><span class="fl">0.05</span>,<span class="at">col=</span><span class="st">&quot;red&quot;</span>)</span></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/hist-random-pvalues-1.png" width="672" /></p>
<p>By comparison, this is the pvalue histogram for our comparison of interest:</p>
<div class="sourceCode" id="cb172"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb172-1"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb172-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(</span>
<span id="cb172-2"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb172-2" aria-hidden="true" tabindex="-1"></a>    tt<span class="sc">$</span>PValue,</span>
<span id="cb172-3"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb172-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">n=</span><span class="dv">100</span>, </span>
<span id="cb172-4"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb172-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">main=</span><span class="st">&quot;BRCA Luminal A vs Basal-Like: nominal pvalues&quot;</span></span>
<span id="cb172-5"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb172-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb172-6"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb172-6" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">v=</span><span class="fl">0.05</span>,<span class="at">col=</span><span class="st">&quot;red&quot;</span>)</span></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/histogram-real-pvalues-1.png" width="672" /></p>
<p>A QQplot directly compares the pvalues from our statistical tests to the expected values from a random uniform distribution:</p>
<div class="sourceCode" id="cb173"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb173-1"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb173-1" aria-hidden="true" tabindex="-1"></a><span class="fu">qqplot</span>(</span>
<span id="cb173-2"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb173-2" aria-hidden="true" tabindex="-1"></a>    tt<span class="sc">$</span>PValue, </span>
<span id="cb173-3"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb173-3" aria-hidden="true" tabindex="-1"></a>    rnd_pval, </span>
<span id="cb173-4"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb173-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">xlab=</span><span class="st">&quot;p-values from real data&quot;</span>,</span>
<span id="cb173-5"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb173-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">ylab=</span><span class="st">&quot;X~U[0,1]&quot;</span>,</span>
<span id="cb173-6"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb173-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">pch=</span><span class="dv">16</span>,<span class="at">cex=</span><span class="fl">0.5</span></span>
<span id="cb173-7"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb173-7" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb173-8"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb173-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb173-9"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb173-9" aria-hidden="true" tabindex="-1"></a><span class="co"># x=y line as reference</span></span>
<span id="cb173-10"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb173-10" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="dv">0</span>,<span class="dv">1</span>,<span class="at">col=</span><span class="st">&quot;red&quot;</span>)</span></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/qqplot-1.png" width="672" /></p>
<p>Now let’s call differentially expressed genes using the <code>decideTestDGE()</code> function and use <code>summary()</code> to see how many genes are upregulated (value <code>+1</code>), downregulated (value <code>-1</code>) and not called as changed (value <code>0</code>)</p>
<div class="sourceCode" id="cb174"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb174-1"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb174-1" aria-hidden="true" tabindex="-1"></a>diffEx2 <span class="ot">&lt;-</span> <span class="fu">decideTestsDGE</span>(diffEx, </span>
<span id="cb174-2"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb174-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">adjust.method=</span><span class="st">&quot;BH&quot;</span>, </span>
<span id="cb174-3"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb174-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">p.value=</span><span class="fl">0.05</span></span>
<span id="cb174-4"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb174-4" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb174-5"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb174-5" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(diffEx2)</span></code></pre></div>
<pre><code>##        groupBasal-like
## Down              2813
## NotSig            1681
## Up                3205</code></pre>
<p>A <strong>volcano plot</strong> can help visualize effect magnitude - log2 fold-change or <code>log2FC</code> in the table ` against the corresponding p-value. Here we create a volcano plot, and colour-code upregulated genes in red, and downregulated genes in blue.</p>
<p>Note that we are combining two different tables, <code>tt</code> and <code>diffEx2</code> so we need to ensure the order is the same. Otherwise the colours will be in the wrong order <em>(try it for yourself!)</em>.</p>
<div class="sourceCode" id="cb176"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb176-1"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb176-1" aria-hidden="true" tabindex="-1"></a>midx <span class="ot">&lt;-</span> <span class="fu">match</span>(<span class="fu">rownames</span>(tt), <span class="fu">rownames</span>(diffEx2))</span>
<span id="cb176-2"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb176-2" aria-hidden="true" tabindex="-1"></a>diffEx2 <span class="ot">&lt;-</span> diffEx2[midx,]</span>
<span id="cb176-3"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb176-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb176-4"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb176-4" aria-hidden="true" tabindex="-1"></a>cols <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="st">&quot;black&quot;</span>,<span class="fu">nrow</span>(diffEx2))</span>
<span id="cb176-5"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb176-5" aria-hidden="true" tabindex="-1"></a>cols[<span class="fu">which</span>(diffEx2<span class="sc">&gt;</span><span class="dv">0</span> )]  <span class="ot">&lt;-</span> <span class="st">&quot;red&quot;</span></span>
<span id="cb176-6"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb176-6" aria-hidden="true" tabindex="-1"></a>cols[<span class="fu">which</span>(diffEx2<span class="sc">&lt;</span><span class="dv">0</span>)]  <span class="ot">&lt;-</span> <span class="st">&quot;blue&quot;</span></span>
<span id="cb176-7"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb176-7" aria-hidden="true" tabindex="-1"></a><span class="co"># volcano plot</span></span>
<span id="cb176-8"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb176-8" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(tt<span class="sc">$</span>logFC,<span class="sc">-</span><span class="fu">log10</span>(tt<span class="sc">$</span>PValue),<span class="at">pch=</span><span class="dv">16</span>,</span>
<span id="cb176-9"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb176-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">col=</span>cols)</span>
<span id="cb176-10"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb176-10" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">v=</span><span class="dv">0</span>,<span class="at">lty=</span><span class="dv">3</span>)</span></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/unnamed-chunk-71-1.png" width="672" /></p>
<p>Finally we can write our differential expression results out to file:</p>
<div class="sourceCode" id="cb177"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb177-1"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb177-1" aria-hidden="true" tabindex="-1"></a><span class="fu">write.table</span>(tt,<span class="at">file=</span><span class="st">&quot;diffEx.results.txt&quot;</span>,</span>
<span id="cb177-2"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb177-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">sep=</span><span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>,</span>
<span id="cb177-3"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb177-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">col=</span><span class="cn">TRUE</span>,</span>
<span id="cb177-4"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb177-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">row=</span><span class="cn">TRUE</span>,</span>
<span id="cb177-5"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb177-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">quote=</span><span class="cn">FALSE</span></span>
<span id="cb177-6"><a href="module-4-finding-differentially-expressed-genes-with-rna-seq.html#cb177-6" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
</div>
<div id="exercise-3" class="section level2 unnumbered">
<h2>Exercise</h2>
<ul>
<li>Install the <code>yeastRNASeq</code> package from Bioconductor and <code>library</code> it into your environment</li>
<li>Import the geneLevelData using: <code>data(yeastRNASeq)</code></li>
<li>Learn about this data and then put it through the same workflow we just did for the breast cancer:</li>
</ul>
<ol style="list-style-type: decimal">
<li>Filter genes with &lt; 2 mean expression across all samples</li>
<li>Create a new expression set object with your filtered genes</li>
<li>Normalize and plot your data</li>
<li>Create a model matrix for analysis</li>
<li>Fit your model</li>
<li>How many significantly up-regulated genes are there at the 5% FDR level? How many significantly down-regulated genes? How many in total</li>
<li>Create a volcano plot</li>
<li>Bonus: Create a histogram of p-values. Is there a signal?</li>
</ol>
<p>Is there anything about the data that might make you question the results?</p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="module-3-exercise.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="module-4---exercises.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": null,
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["bookdown-demo.pdf", "bookdown-demo.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
